File: 841-KeysandRooms/841-KeysandRooms.java
/*

841. Keys and Rooms
Medium

There are N rooms and you start in room 0.  Each room has a distinct number in 0, 1, 2, ..., N-1, and each room may have some keys to access the next room. 

Formally, each room i has a list of keys rooms[i], and each key rooms[i][j] is an integer in [0, 1, ..., N-1] where N = rooms.length.  A key rooms[i][j] = v opens the room with number v.

Initially, all the rooms start locked (except for room 0). 

You can walk back and forth between rooms freely.

Return true if and only if you can enter every room.

Example 1:

Input: [[1],[2],[3],[]]
Output: true
Explanation:  
We start in room 0, and pick up key 1.
We then go to room 1, and pick up key 2.
We then go to room 2, and pick up key 3.
We then go to room 3.  Since we were able to go to every room, we return true.

Example 2:

Input: [[1,3],[3,0,1],[2],[0]]
Output: false
Explanation: We can't enter the room with number 2.

Note:

    1 <= rooms.length <= 1000
    0 <= rooms[i].length <= 1000
    The number of keys in all rooms combined is at most 3000.

*/

class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        Map<Integer, List<Integer>> graph = getGraph(rooms);
        Set<Integer> visited = new HashSet<>();
        dfs(graph, 0, visited);
        return visited.size() == graph.keySet().size();
    }
    
    private void dfs(Map<Integer, List<Integer>> graph , int r, Set<Integer> visited){
        if(visited.contains(r)) return;
        visited.add(r);
        for(int room : graph.get(r)){
            dfs(graph, room, visited);
        }
    }
    
    private Map<Integer, List<Integer>> getGraph(List<List<Integer>> rooms){
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int i = 0; i < rooms.size(); i++) graph.put(i, new ArrayList<>());
        for(int i = 0; i < rooms.size(); i++){
            for(int j = 0; j < rooms.get(i).size(); j++){
                graph.get(i).add(rooms.get(i).get(j));
            }
        }
        return graph;
    }
}

File: 841-KeysandRooms/841-KeysandRooms.py
"""

841. Keys and Rooms

There are n rooms labeled from 0 to n - 1 and all the rooms are locked except for room 0. Your goal is to visit all the rooms. However, you cannot enter a locked room without having its key.

When you visit a room, you may find a set of distinct keys in it. Each key has a number on it, denoting which room it unlocks, and you can take all of them with you to unlock the other rooms.

Given an array rooms where rooms[i] is the set of keys that you can obtain if you visited room i, return true if you can visit all the rooms, or false otherwise.

Example 1:

Input: rooms = [[1],[2],[3],[]]
Output: true
Explanation: 
We visit room 0 and pick up key 1.
We then visit room 1 and pick up key 2.
We then visit room 2 and pick up key 3.
We then visit room 3.
Since we were able to visit every room, we return true.

Example 2:

Input: rooms = [[1,3],[3,0,1],[2],[0]]
Output: false
Explanation: We can not enter room number 2 since the only key that unlocks it is in that room.

Constraints:

    n == rooms.length
    2 <= n <= 1000
    0 <= rooms[i].length <= 1000
    1 <= sum(rooms[i].length) <= 3000
    0 <= rooms[i][j] < n
    All the values of rooms[i] are unique.

"""

class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        def build_graph():
            graph = defaultdict(list)
            for i in range(len(rooms)):
                graph[i].extend(rooms[i])
            return graph
        graph = build_graph()
        def dfs(idx):
            if vis[idx] == True: return
            vis[idx] = True
            res[0] += 1
            for child in graph[idx]:
                dfs(child)
        vis = [False] * len(rooms)
        res = [0]
        dfs(0)
        return res[0] == len(rooms)
